src/daft-writers/src/partition.rs

```diff
@@ -0,0 +1,148 @@
+use std::{
+    collections::{hash_map::RawEntryMut, HashMap},
+    sync::Arc,
+};
+
+use common_error::DaftResult;
+use daft_core::{array::ops::as_arrow::AsArrow, utils::identity_hash_set::IndexHash};
+use daft_dsl::ExprRef;
+use daft_io::IOStatsContext;
+use daft_micropartition::MicroPartition;
+use daft_table::Table;
+
+use crate::{FileWriter, WriterFactory};
+
+/// PartitionedWriter is a writer that partitions the input data by a set of columns, and writes each partition
+/// to a separate file. It uses a map to keep track of the writers for each partition.
+struct PartitionedWriter {
+    per_partition_writers:
+        HashMap<IndexHash, Box<dyn FileWriter<Input = Arc<MicroPartition>, Result = Vec<Table>>>>,
```

Comment:
tbh, we should do something like a `ScalarValue` that wraps the possible partition value types for our non performance critical stuff like here.